[onewire] - Only tries initial owserver connection once?

OpenHAB 1.8.3 with onewire binding active appears to only attempt a connection to the designated owserver server (in this case at 127.0.0.1) ONCE, when OpenHAB initially starts up.

If the owserver daemon either isn't ready & listening when OpenHAB starts, the OH onewire binding --> owserver TCP connection fails, and is never retried. 

Similarly, if the owserver daemon needs to be restarted at some later time, perhaps because it quit of its own accord or as part of other system maintenance, OpenHAB & onewire binding never again look for the owserver.

## Expected Behavior
Suggested behavior:  Make available in the onewire section of openhab.cfg a configuration option defining how frequently the onewire binding should re-attempt to retry and/or reset a broken or non-existent initial owserver TCP connection.

## Current Behavior
When openhab.sh is first started, when the onewire binding loads it attempt to connect to the defined owserver IP / port. (In my case, 127.0.0.1:4304).  If the owserver is not responding or is not started, openhab.sh & onewire binding never again try to locate the owserver. It is necessary to kill & restart openhab.sh to re-establish the connection.

## Your Environment
* Version used: OH 1.8.3
* Environment name and version: Java(TM) SE Runtime Environment (build 1.8.0_91-b14)
* Operating System and version: Ubuntu 16.04 LTS / i386

Tested.
Works fine in both these scenarios:
1) owserver already up & running when OpenHAB is started up, then owserver killed, and manually restarted.

2) owserver NOT up & running when OpenHAB is started. (owserver manually started up after OH is already running.)

Comments:
The retrying of individual sensors is fundamentally different from the retrying of the owserver:4304 TCP connection, so I do not think the 2 functions should be bundled into 1 config option.  It makes sense to try a flaky sensor X number of times, then give up (at least until the next polling loop) -- in fact it is necessary to give up on a sensor at some point, else the binding never moves on to the healthy sensors in line after the bad one(s).

But the owserver:4304 TCP connection should be retried at a specified interval indefinitely, because an OH setup that depends on 1Wire sensor data (e.g. temperature for example) will not function correctly without that data.  

I agree that 5 seconds is probably fine as a minimum interval, that gives good responsiveness while keeping CPU/network load reasonable.

My comments are inserted into the log output, marked with ### below:
Log file output:
`

### owserver daemon was up and running at time OpenHAB was started: ---------
###

20:10:29.101 [INFO ] [o.o.b.o.i.c.OneWireConnection :150  ] - Connected to owserver [IP 'localhost' Port '4304']
20:10:30.108 [INFO ] [runtime.busevents             :26   ] - BackBedTempSensor state updated to 59.78750000000000142108547152020037174224853515625
20:10:31.228 [INFO ] [runtime.busevents             :26   ] - KitchenTempSensor state updated to 59.78750000000000142108547152020037174224853515625
20:10:33.517 [INFO ] [runtime.busevents             :26   ] - FrontBedTempSensor0 state updated to 61.47500000000000142108547152020037174224853515625

20:10:34.863 [DEBUG] [m.r.internal.engine.RuleEngine:305  ] - Executing rule 'FRONT BEDROOM THERMOSTATIC CONTROL'
20:10:34.879 [INFO ] [runtime.busevents             :26   ] - FrontBedroomTempSensor state updated to 63.72500000000000142108547152020037174224853515625
20:11:09.732 [INFO ] [runtime.busevents             :26   ] - FrontBedTempSensor5 state updated to 65.525000000000005684341886080801486968994140625#

###
###MANUALLY KILLED -9'ed the owserver daemon ----------
###

20:12:29.114 [ERROR] [o.o.b.o.i.c.OneWireConnection :299  ] - couldn't establish network connection while read attempt 1 '28.FFC7B870160520/temperature' ip:port=localhost:4304
java.net.SocketException: Broken pipe
	at java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:1.8.0_91]
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109) ~[na:1.8.0_91]
	at java.net.SocketOutputStream.write(SocketOutputStream.java:132) ~[na:1.8.0_91]
	at java.io.DataOutputStream.writeInt(DataOutputStream.java:198) ~[na:1.8.0_91]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.sendRequest(OwfsConnectionImpl.java:209) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.exists(OwfsConnectionImpl.java:158) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.openhab.binding.onewire.internal.connection.OneWireConnection.checkIfDeviceExists(OneWireConnection.java:251) ~[bundlefile:na]
	at org.openhab.binding.onewire.internal.connection.OneWireConnection.readFromOneWire(OneWireConnection.java:268) ~[bundlefile:na]
	at org.openhab.binding.onewire.internal.OneWireBinding.updateItemFromOneWire(OneWireBinding.java:293) [bundlefile:na]
	at org.openhab.binding.onewire.internal.OneWireBinding.devicePropertyWantsUpdate(OneWireBinding.java:247) [bundlefile:na]
	at org.openhab.binding.onewire.internal.scheduler.OneWireUpdateTask.run(OneWireUpdateTask.java:61) [bundlefile:na]

20:12:29.118 [INFO ] [o.o.b.o.i.c.OneWireConnection :161  ] - Trying to reconnect to owserver...
20:12:29.121 [WARN ] [o.o.b.o.i.c.OneWireConnection :116  ] - Unexpected connection failure.
java.net.ConnectException: Connection refused
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_91]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_91]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_91]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_91]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_91]
	at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_91]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.tryToSocketConnectionAndStreamsInitialization(OwfsConnectionImpl.java:79) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.connect(OwfsConnectionImpl.java:70) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.establishConnectionIfNeeded(OwfsConnectionImpl.java:113) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.sendRequest(OwfsConnectionImpl.java:204) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.owfs.jowfsclient.internal.OwfsConnectionImpl.listDirectory(OwfsConnectionImpl.java:193) ~[org.owfs.jowfsclient-1.2.6.jar:na]
	at org.openhab.binding.onewire.internal.connection.OneWireConnection.connect(OneWireConnection.java:108) [bundlefile:na]
	at org.openhab.binding.onewire.internal.connection.OneWireConnection.reconnect(OneWireConnection.java:168) [bundlefile:na]
	at org.openhab.binding.onewire.internal.connection.OneWireConnection.readFromOneWire(OneWireConnection.java:301) [bundlefile:na]
	at org.openhab.binding.onewire.internal.OneWireBinding.updateItemFromOneWire(OneWireBinding.java:293) [bundlefile:na]
	at org.openhab.binding.onewire.internal.OneWireBinding.devicePropertyWantsUpdate(OneWireBinding.java:247) [bundlefile:na]
	at org.openhab.binding.onewire.internal.scheduler.OneWireUpdateTask.run(OneWireUpdateTask.java:61) [bundlefile:na]

20:12:29.125 [WARN ] [o.o.b.o.i.c.OneWireConnection :120  ] - Connection failed. Will retry in 60 seconds.

20:13:26.514 [INFO ] [runtime.busevents             :26   ] - Date state updated to 2016-12-04T20:13:26
20:13:29.126 [INFO ] [o.o.b.o.i.c.OneWireConnection :130  ] - Retrying failed connection...
20:13:29.129 [WARN ] [o.o.b.o.i.c.OneWireConnection :140  ] - Unexpected connection failure.

###deleted repeated "java.net.PlainSocket" & following error lines (from above) to save space---)
###

###
### MANUALLY RESTARTED owserver daemon (while OpenHAB was still running -----
###

20:14:29.133 [INFO ] [o.o.b.o.i.c.OneWireConnection :130  ] - Retrying failed connection...
20:14:30.381 [INFO ] [o.o.b.o.i.c.OneWireConnection :150  ] - Connected to owserver [IP 'localhost' Port '4304']
20:14:31.635 [INFO ] [runtime.busevents             :26   ] - BackBedTempSensor state updated to 59.6749999999999971578290569595992565155029296875
`

--- a/bundles/binding/org.openhab.binding.onewire/META-INF/MANIFEST.MF
+++ b/bundles/binding/org.openhab.binding.onewire/META-INF/MANIFEST.MF
@@ -30,6 +30,7 @@ Bundle-SymbolicName: org.openhab.binding.onewire
 Bundle-DocURL: http://www.openhab.org
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Service-Component: OSGI-INF/binding.xml, OSGI-INF/genericbindingprovider.xml
+Bundle-ActivationPolicy: lazy
 Bundle-ClassPath: .,
  lib/org.owfs.jowfsclient-1.2.6.jar
 

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireBinding.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireBinding.java
@@ -112,7 +112,7 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
 
     @Override
     protected void internalReceiveCommand(String pvItemName, Command pvCommand) {
-        logger.debug("received command {} for item {}", pvCommand.toString(), pvItemName);
+        logger.debug("received command {} for item {}", pvCommand, pvItemName);
 
         OneWireBindingConfig lvBindigConfig = getBindingConfig(pvItemName);
 
@@ -126,7 +126,7 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
             AbstractOneWireControlBindingConfig lvControlBindingConfig = (AbstractOneWireControlBindingConfig) lvBindigConfig;
             lvControlBindingConfig.executeControl(this, pvCommand);
         } else {
-            logger.debug("received command {} for item {} which is not writable or executable", pvCommand.toString(),
+            logger.debug("received command {} for item {} which is not writable or executable", pvCommand,
                     pvItemName);
         }
     }
@@ -224,7 +224,7 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
                     }
                 } else {
                     logger.debug("Didn't add to OneWireUpdate scheduler, because refresh is <= 0: {}",
-                            lvDeviceBindingConfig.toString());
+                            lvDeviceBindingConfig);
                 }
             }
         }

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireBinding.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireBinding.java
@@ -112,7 +112,7 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
 
     @Override
     protected void internalReceiveCommand(String pvItemName, Command pvCommand) {
-        logger.debug("received command " + pvCommand.toString() + " for item " + pvItemName);
+        logger.debug("received command {} for item {}", pvCommand.toString(), pvItemName);
 
         OneWireBindingConfig lvBindigConfig = getBindingConfig(pvItemName);
 
@@ -126,8 +126,8 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
             AbstractOneWireControlBindingConfig lvControlBindingConfig = (AbstractOneWireControlBindingConfig) lvBindigConfig;
             lvControlBindingConfig.executeControl(this, pvCommand);
         } else {
-            logger.debug("received command " + pvCommand.toString() + " for item " + pvItemName
-                    + " which is not writable or executable");
+            logger.debug("received command {} for item {} which is not writable or executable", pvCommand.toString(),
+                    pvItemName);
         }
     }
 
@@ -223,8 +223,8 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
                         logger.warn("Couldn't add to OneWireUpdate scheduler", lvDeviceBindingConfig);
                     }
                 } else {
-                    logger.debug("Didn't add to OneWireUpdate scheduler, because refresh is <= 0: "
-                            + lvDeviceBindingConfig.toString());
+                    logger.debug("Didn't add to OneWireUpdate scheduler, because refresh is <= 0: {}",
+                            lvDeviceBindingConfig.toString());
                 }
             }
         }
@@ -242,7 +242,7 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
     public void devicePropertyWantsUpdate(OneWireDevicePropertyWantsUpdateEvent pvWantsUpdateEvent) {
         String lvItemName = pvWantsUpdateEvent.getItemName();
 
-        logger.debug("Item " + lvItemName + " wants update!");
+        logger.debug("Item {} wants update!", lvItemName);
 
         updateItemFromOneWire(lvItemName);
     }
@@ -284,8 +284,8 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
                     pvItemName);
 
             if (pvBindingConfig == null) {
-                logger.error("no bindingConfig found for itemName=" + pvItemName
-                        + " cannot update! It will be removed from scheduler");
+                logger.error("no bindingConfig found for itemName={} cannot update! It will be removed from scheduler",
+                        pvItemName);
                 ivOneWireReaderScheduler.removeItem(pvItemName);
                 return;
             }
@@ -298,7 +298,7 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
                 if (lvItem != null) {
                     postUpdate(lvItem, lvNewType);
                 } else {
-                    logger.error("There is no Item for ItemName=" + pvItemName);
+                    logger.error("There is no Item for ItemName={}", pvItemName);
                 }
             } else {
                 String lvLogText = "Set Item for itemName=" + pvItemName
@@ -323,8 +323,8 @@ public class OneWireBinding extends AbstractBinding<OneWireBindingProvider>
                 ivCacheItemStates.put(pvItem.getName(), lvNewState);
                 eventPublisher.postUpdate(pvItem.getName(), lvNewState);
             } else {
-                logger.debug("didn't post update to eventPublisher, because state did not change for item "
-                        + pvItem.getName());
+                logger.debug("didn't post update to eventPublisher, because state did not change for item {}",
+                        pvItem.getName());
             }
         }
     }

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireBindingConfigFactory.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireBindingConfigFactory.java
@@ -42,7 +42,7 @@ public class OneWireBindingConfigFactory {
      */
     public static OneWireBindingConfig createOneWireDeviceProperty(Item pvItem, String pvBindingConfig)
             throws BindingConfigParseException {
-        logger.debug("createOneWireDeviceProperty: " + pvItem.getName() + " - bindingConfig:" + pvBindingConfig);
+        logger.debug("createOneWireDeviceProperty: {} - bindingConfig:{}", pvItem.getName(), pvBindingConfig);
 
         OneWireBindingConfig lvNewBindingConfig = null;
         if (OneWireClearCacheControlBindingConfig.isBindingConfigToCreate(pvItem, pvBindingConfig)) {
@@ -63,7 +63,7 @@ public class OneWireBindingConfigFactory {
                     "the item-type " + pvItem.getClass() + " cannot be a onewire device");
         }
 
-        logger.debug("created newBindingConfig: " + lvNewBindingConfig.toString());
+        logger.debug("created newBindingConfig: {}", lvNewBindingConfig.toString());
 
         return lvNewBindingConfig;
     }
@@ -80,8 +80,8 @@ public class OneWireBindingConfigFactory {
                 || (pvItem instanceof SwitchItem) || (pvItem instanceof StringItem));
 
         if (!lvIsValidItem) {
-            logger.error("Item " + pvItem.getName() + " of type " + pvItem.getClass().getSimpleName()
-                    + " with configuration " + pvBindingConfig + " is not a valid onewire ItemType!");
+            logger.error("Item {} of type {} with configuration {} is not a valid onewire ItemType!", pvItem.getName(),
+                    pvItem.getClass().getSimpleName(), pvBindingConfig);
         }
 
         return lvIsValidItem;

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireGenericBindingProvider.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OneWireGenericBindingProvider.java
@@ -54,7 +54,7 @@ public class OneWireGenericBindingProvider extends AbstractGenericBindingProvide
      */
     @Override
     public void validateItemType(Item pvItem, String pvBindingConfig) throws BindingConfigParseException {
-        logger.debug("validateItemType: " + pvItem.getName() + " - bindingConfig:" + pvBindingConfig);
+        logger.debug("validateItemType: {} - bindingConfig:{}", pvItem.getName(), pvBindingConfig);
 
         if (OneWireBindingConfigFactory.isValidItemType(pvItem, pvBindingConfig)) {
             return;

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/connection/OneWireConnection.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/connection/OneWireConnection.java
@@ -178,7 +178,7 @@ public class OneWireConnection {
 
     public static synchronized void updated(Dictionary<String, ?> pvConfig) throws ConfigurationException {
         if (pvConfig == null) {
-            logger.info(
+            logger.debug(
                     "OneWireBinding configuration is not present. Please check your configuration file or if not needed remove the OneWireBinding addon.");
             return;
         }
@@ -341,13 +341,13 @@ public class OneWireConnection {
         while (lvAttempt <= cvRetry) {
             try {
                 logger.debug("Trying to write '{}' to '{}', write attempt={}",
-                        new Object[] { pvValue, pvDevicePropertyPath, lvAttempt });
+                        pvValue, pvDevicePropertyPath, lvAttempt);
                 if (checkIfDeviceExists(pvDevicePropertyPath)) {
                     OneWireConnection.getConnection().write(pvDevicePropertyPath, pvValue);
                     return; // Success, exit
                 } else {
                     logger.info("There is no device for path {}, write attempt={}",
-                            new Object[] { pvDevicePropertyPath, lvAttempt });
+                            pvDevicePropertyPath, lvAttempt);
                 }
             } catch (OwfsException oe) {
                 logger.error("Writing {} to path {} attempt {} threw an exception", pvValue, pvDevicePropertyPath,

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/connection/OneWireConnection.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/connection/OneWireConnection.java
@@ -29,6 +29,7 @@ import org.slf4j.LoggerFactory;
  * This class establishes the connection to the 1-Wire-bus.
  *
  * @author Dennis Riegelbauer
+ * @author Chris Carman (added server connection retry logic)
  * @since 1.7.0
  *
  */

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/connection/OneWireConnection.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/connection/OneWireConnection.java
@@ -10,6 +10,7 @@ package org.openhab.binding.onewire.internal.connection;
 
 import java.io.IOException;
 import java.util.Dictionary;
+import java.util.List;
 
 import org.apache.commons.lang.StringUtils;
 import org.openhab.binding.onewire.internal.deviceproperties.AbstractOneWireDevicePropertyBindingConfig;
@@ -61,13 +62,25 @@ public class OneWireConnection {
     private static int cvRetry = 3;
 
     /**
+     * The number of retries that will be attempted after a failed connection attempt.
+     * Optional, defaults to 3. 0 means no retries will be attempted.
+     */
+    private static int cvServerRetries = 3;
+
+    /**
+     * The time to wait between connection attempts. Optional, defaults to 60 seconds.
+     * May not be less than 5 seconds.
+     */
+    private static int cvServerRetryInterval = 60;
+
+    /**
      * signals that the connection is established
      */
     private static boolean cvIsEstablished = false;
 
     /**
-     * Returns a OwfsConnection
-     * 
+     * Returns an OwfsConnection
+     *
      * @return the OwfsConnection network link
      */
     public static synchronized OwfsConnection getConnection() {
@@ -81,40 +94,74 @@ public class OneWireConnection {
 
     /**
      * Tries to connect either by IP or serial bus, depending on supplied config data.
-     * 
+     *
      * @return true if connection was established, false otherwise
      */
     public static synchronized boolean connect() {
-        if (cvIp != null && cvPort > 0) {
-            OwfsConnectionFactory owfsConnectorFactory = new OwfsConnectionFactory(cvIp, cvPort);
-            OwfsConnectionConfig owConnectionConfig = new OwfsConnectionConfig(cvIp, cvPort);
-            owConnectionConfig.setTemperatureScale(cvTempScale);
-            owConnectionConfig.setPersistence(OwPersistence.ON);
-            owConnectionConfig.setBusReturn(OwBusReturn.ON);
-            owfsConnectorFactory.setConnectionConfig(owConnectionConfig);
+        OwfsConnectionFactory owfsConnectorFactory = new OwfsConnectionFactory(cvIp, cvPort);
+        OwfsConnectionConfig owConnectionConfig = new OwfsConnectionConfig(cvIp, cvPort);
+        owConnectionConfig.setTemperatureScale(cvTempScale);
+        owConnectionConfig.setPersistence(OwPersistence.ON);
+        owConnectionConfig.setBusReturn(OwBusReturn.ON);
+        owfsConnectorFactory.setConnectionConfig(owConnectionConfig);
 
-            try {
-                cvOwConnection = owfsConnectorFactory.createNewConnection();
-                cvOwConnection.listDirectory("/");
-                logger.info("Connected to owserver [IP '" + cvIp + "' Port '" + cvPort + "']");
-                cvIsEstablished = true;
-                return true;
-            } catch (Exception exception) {
-                logger.error("Couldn't connect to owserver [IP '" + cvIp + "' Port '" + cvPort + "']: ",
-                        exception.getLocalizedMessage());
+        cvOwConnection = owfsConnectorFactory.createNewConnection();
+
+        boolean connected = false;
+        int attempts = 0, retriesRemaining = cvServerRetries;
+        List<String> result = null;
+
+        try {
+            result = cvOwConnection.listDirectory("/");
+            if (result != null) {
+                connected = true;
+            } else {
                 cvIsEstablished = false;
-                return false;
             }
-        } else {
-            logger.warn("Couldn't connect to owserver because of missing connection parameters [IP '{}' Port '{}'].",
-                    cvIp, cvPort);
+        } catch (OwfsException oe) {
+            logger.warn("Unexpected owfs exception: {}", oe.getMessage(), oe);
+        } catch (IOException e) {
+            logger.warn("Unexpected connection failure.", e);
+        }
+
+        while (!connected && retriesRemaining > 0) {
+            logger.warn("Connection failed. Will retry in {} seconds.", cvServerRetryInterval);
+            synchronized (cvOwConnection) {
+                try {
+                    cvOwConnection.wait(cvServerRetryInterval * 1000L);
+                } catch (InterruptedException e) {
+                    logger.debug("Wait was interrupted.");
+                }
+            }
+            attempts++;
+            retriesRemaining--;
+            logger.info("Retrying failed connection... Attempt {} of {}.", attempts, cvServerRetries);
+            try {
+                result = cvOwConnection.listDirectory("/");
+                if (result != null) {
+                    connected = true;
+                }
+            } catch (OwfsException oe) {
+                logger.warn("Unexpected owfs exception: {}", oe.getMessage(), oe);
+            } catch (IOException e) {
+                logger.warn("Unexpected connection failure.", e);
+            }
+        }
+
+        if (!connected) {
+            logger.error("Couldn't connect to owserver [IP '{}' Port '{}']", cvIp, cvPort);
+            cvIsEstablished = false;
             return false;
         }
+
+        logger.info("Connected to owserver [IP '{}' Port '{}']", cvIp, cvPort);
+        cvIsEstablished = true;
+        return true;
     }
 
     /**
      * Reconnects to owserver
-     * 
+     *
      * @return
      */
     public static synchronized boolean reconnect() {
@@ -122,50 +169,80 @@ public class OneWireConnection {
         try {
             cvOwConnection.disconnect();
         } catch (Exception lvException) {
-            logger.error("Error while disconnecting from owserver: " + lvException, lvException);
+            logger.error("Error while disconnecting from owserver: ", lvException);
         }
         cvOwConnection = null;
+        cvIsEstablished = false;
         return connect();
     }
 
     public static synchronized void updated(Dictionary<String, ?> pvConfig) throws ConfigurationException {
-        if (pvConfig != null) {
-            logger.debug("OneWire configuration present. Setting up owserver connection.");
-            cvIp = (String) pvConfig.get("ip");
-
-            String lvPortConfig = (String) pvConfig.get("port");
-            if (StringUtils.isNotBlank(lvPortConfig)) {
-                cvPort = Integer.parseInt(lvPortConfig);
-            }
-
-            String lvTempScaleString = (String) pvConfig.get("tempscale");
-            if (StringUtils.isNotBlank(lvTempScaleString)) {
-                try {
-                    cvTempScale = OwTemperatureScale.valueOf(lvTempScaleString);
-                } catch (IllegalArgumentException iae) {
-                    String lvFehlertext = "Unknown temperature scale '" + lvTempScaleString
-                            + "'. Valid values are CELSIUS, FAHRENHEIT, KELVIN or RANKINE.";
-                    logger.error(lvFehlertext, iae);
-                    throw new ConfigurationException("onewire:tempscale", lvFehlertext);
-                }
-            }
-
-            String lvRetryString = (String) pvConfig.get("retry");
-            if (StringUtils.isNotBlank(lvRetryString)) {
-                cvRetry = Integer.parseInt(lvRetryString);
-            }
-
-            if (cvOwConnection == null) {
-                logger.debug("Not connected to owserver yet. Trying to connect...");
-                if (!connect()) {
-                    logger.warn("Initial connection to owserver failed!");
-                } else {
-                    logger.debug("Success: connected to owserver.");
-                }
-            }
-        } else {
+        if (pvConfig == null) {
             logger.info(
                     "OneWireBinding configuration is not present. Please check your configuration file or if not needed remove the OneWireBinding addon.");
+            return;
+        }
+
+        logger.debug("OneWire configuration present. Setting up owserver connection.");
+        cvIp = (String) pvConfig.get("ip");
+        if (StringUtils.isBlank(cvIp)) {
+            logger.error("owserver IP address was configured as an empty string.");
+            throw new ConfigurationException("onewire:ip", "owserver IP address was configured as an empty string.");
+        }
+
+        String lvPortConfig = (String) pvConfig.get("port");
+        if (StringUtils.isNotBlank(lvPortConfig)) {
+            cvPort = Integer.parseInt(lvPortConfig);
+        }
+        if (cvPort < 1) {
+            logger.error("owserver port was configured with an invalid value: {}", cvPort);
+            throw new ConfigurationException("onewire:port",
+                    "owserver port was configured with an invalid value: " + cvPort);
+        }
+        logger.debug("owserver ip:port = {}:{}", cvIp, cvPort);
+
+        String lvTempScaleString = (String) pvConfig.get("tempscale");
+        if (StringUtils.isNotBlank(lvTempScaleString)) {
+            try {
+                cvTempScale = OwTemperatureScale.valueOf(lvTempScaleString);
+            } catch (IllegalArgumentException iae) {
+                String lvFehlertext = "Unknown temperature scale '" + lvTempScaleString
+                        + "'. Valid values are CELSIUS, FAHRENHEIT, KELVIN or RANKINE.";
+                logger.error(lvFehlertext, iae);
+                throw new ConfigurationException("onewire:tempscale", lvFehlertext);
+            }
+        }
+
+        String lvRetryString = (String) pvConfig.get("retry");
+        if (StringUtils.isNotBlank(lvRetryString)) {
+            cvRetry = Integer.parseInt(lvRetryString);
+        }
+        logger.debug("onewire:retry = {}", cvRetry);
+
+        String lvServerRetries = (String) pvConfig.get("server_retries");
+        if (StringUtils.isNotBlank(lvServerRetries)) {
+            cvServerRetries = Integer.parseInt(lvServerRetries);
+        }
+        logger.debug("onewire:server_retries = {}", cvServerRetries);
+
+        String lvRetryIntervalString = (String) pvConfig.get("server_retryInterval");
+        if (StringUtils.isNotBlank(lvRetryIntervalString)) {
+            cvServerRetryInterval = Integer.parseInt(lvRetryIntervalString);
+            if (cvServerRetryInterval < 5 && cvServerRetryInterval > 0) {
+                logger.info("server_retryInterval was set to {}. Using the minimum allowed value of 5 instead.",
+                        cvServerRetryInterval);
+                cvServerRetryInterval = 5;
+            }
+        }
+        logger.debug("onewire:server_retryInterval = {} seconds", cvServerRetryInterval);
+
+        if (cvOwConnection == null) {
+            logger.debug("Not connected to owserver yet. Trying to connect...");
+            if (!connect()) {
+                logger.warn("Connection to owserver failed!");
+            } else {
+                logger.debug("Success: connected to owserver.");
+            }
         }
     }
 
@@ -178,7 +255,7 @@ public class OneWireConnection {
 
     /**
      * Checks if an device exists in 1-Wire network
-     * 
+     *
      * @param pvDevicePropertyPath
      * @return
      * @throws IOException
@@ -196,7 +273,7 @@ public class OneWireConnection {
 
     /**
      * Read a Value for a device property from 1-Wire network
-     * 
+     *
      * @param pvDevicePropertyPath
      * @return device property value as String
      */
@@ -217,8 +294,8 @@ public class OneWireConnection {
                     if (pvBindingConfig.isIgnore85CPowerOnResetValues()) {
                         double lvReadDouble = Double.parseDouble(lvReadValue);
                         if (lvReadDouble == 85.0) {
-                            logger.debug("reading from path " + lvDevicePropertyPath + " attempt " + lvAttempt
-                                    + " Ignoring 85°C value");
+                            logger.debug("reading from path '{}' attempt {}. Ignoring 85C value", lvDevicePropertyPath,
+                                    lvAttempt);
                         } else {
                             return lvReadValue;
                         }
@@ -239,8 +316,8 @@ public class OneWireConnection {
                     reconnect();
                 }
             } catch (IOException ioe) {
-                logger.error("couldn't establish network connection while read attempt " + lvAttempt + " '"
-                        + lvDevicePropertyPath + "' ip:port=" + cvIp + ":" + cvPort, ioe);
+                logger.error("couldn't establish network connection while read attempt {} '{}'" + " ip:port={}:{}",
+                        lvAttempt, lvDevicePropertyPath, cvIp, cvPort, ioe);
                 reconnect();
             } catch (NumberFormatException lvNumberFormatException) {
                 logger.error(
@@ -255,7 +332,7 @@ public class OneWireConnection {
 
     /**
      * Writes String to 1-Wire device property
-     * 
+     *
      * @param pvDevicePropertyPath
      * @param pvValue
      */
@@ -263,22 +340,22 @@ public class OneWireConnection {
         int lvAttempt = 1;
         while (lvAttempt <= cvRetry) {
             try {
-                logger.debug("trying to write '{}' to '{}', write attempt={}",
+                logger.debug("Trying to write '{}' to '{}', write attempt={}",
                         new Object[] { pvValue, pvDevicePropertyPath, lvAttempt });
                 if (checkIfDeviceExists(pvDevicePropertyPath)) {
                     OneWireConnection.getConnection().write(pvDevicePropertyPath, pvValue);
                     return; // Success, exit
                 } else {
-                    logger.info("there is no device for path {}, write attempt={}",
+                    logger.info("There is no device for path {}, write attempt={}",
                             new Object[] { pvDevicePropertyPath, lvAttempt });
                 }
             } catch (OwfsException oe) {
-                logger.error("writing " + pvValue + " to path " + pvDevicePropertyPath + " attempt " + lvAttempt
-                        + " throws exception", oe);
+                logger.error("Writing {} to path {} attempt {} threw an exception", pvValue, pvDevicePropertyPath,
+                        lvAttempt, oe);
                 reconnect();
             } catch (IOException ioe) {
-                logger.error("couldn't establish network connection while write attempt " + lvAttempt + " to '"
-                        + pvDevicePropertyPath + "' ip:port=" + cvIp + ":" + cvPort, ioe);
+                logger.error("Couldn't establish network connection while write attempt {} to '{}'" + " ip:port={}:{}",
+                        lvAttempt, pvDevicePropertyPath, cvIp, cvPort, ioe);
                 reconnect();
             } finally {
                 lvAttempt++;

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyBindingConfig.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyBindingConfig.java
@@ -210,11 +210,11 @@ public abstract class AbstractOneWireDevicePropertyBindingConfig implements OneW
         Type lvType = convertReadValueToUnmodifiedType(pvReadValue);
 
         for (InterfaceOneWireTypeModifier lvTypeModifier : getTypeModifieryList()) {
-            logger.debug("type of " + getDevicePropertyPath() + " before modifier:" + lvTypeModifier.getModifierName()
-                    + "type=" + lvType.toString());
+            logger.debug("type of {} before modifier:{} type={}", getDevicePropertyPath(),
+                    lvTypeModifier.getModifierName(), lvType.toString());
             lvType = lvTypeModifier.modify4Read(lvType);
-            logger.debug("type of " + getDevicePropertyPath() + " after modifier:" + lvTypeModifier.getModifierName()
-                    + "type=" + lvType.toString());
+            logger.debug("type of {} after modifier:{} type={}", getDevicePropertyPath(),
+                    lvTypeModifier.getModifierName(), lvType.toString());
         }
 
         return lvType;

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyBindingConfig.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyBindingConfig.java
@@ -211,10 +211,10 @@ public abstract class AbstractOneWireDevicePropertyBindingConfig implements OneW
 
         for (InterfaceOneWireTypeModifier lvTypeModifier : getTypeModifieryList()) {
             logger.debug("type of {} before modifier:{} type={}", getDevicePropertyPath(),
-                    lvTypeModifier.getModifierName(), lvType.toString());
+                    lvTypeModifier.getModifierName(), lvType);
             lvType = lvTypeModifier.modify4Read(lvType);
             logger.debug("type of {} after modifier:{} type={}", getDevicePropertyPath(),
-                    lvTypeModifier.getModifierName(), lvType.toString());
+                    lvTypeModifier.getModifierName(), lvType);
         }
 
         return lvType;

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyWritableBindingConfig.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyWritableBindingConfig.java
@@ -39,10 +39,10 @@ public abstract class AbstractOneWireDevicePropertyWritableBindingConfig
     public String convertTypeToString(Type pvType) {
         for (InterfaceOneWireTypeModifier lvTypeModifier : getTypeModifieryList()) {
             logger.debug("type of {} before modifier:{} type={}", getDevicePropertyPath(),
-                    lvTypeModifier.getModifierName(), pvType.toString());
+                    lvTypeModifier.getModifierName(), pvType);
             pvType = lvTypeModifier.modify4Write(pvType);
             logger.debug("type of {} after modifier:{} type={}", getDevicePropertyPath(),
-                    lvTypeModifier.getModifierName(), pvType.toString());
+                    lvTypeModifier.getModifierName(), pvType);
         }
 
         return convertTypeToUnmodifiedString(pvType);

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyWritableBindingConfig.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/deviceproperties/AbstractOneWireDevicePropertyWritableBindingConfig.java
@@ -38,11 +38,11 @@ public abstract class AbstractOneWireDevicePropertyWritableBindingConfig
      */
     public String convertTypeToString(Type pvType) {
         for (InterfaceOneWireTypeModifier lvTypeModifier : getTypeModifieryList()) {
-            logger.debug("type of " + getDevicePropertyPath() + " before modifier:" + lvTypeModifier.getModifierName()
-                    + "type=" + pvType.toString());
+            logger.debug("type of {} before modifier:{} type={}", getDevicePropertyPath(),
+                    lvTypeModifier.getModifierName(), pvType.toString());
             pvType = lvTypeModifier.modify4Write(pvType);
-            logger.debug("type of " + getDevicePropertyPath() + " after modifier:" + lvTypeModifier.getModifierName()
-                    + "state=" + pvType.toString());
+            logger.debug("type of {} after modifier:{} type={}", getDevicePropertyPath(),
+                    lvTypeModifier.getModifierName(), pvType.toString());
         }
 
         return convertTypeToUnmodifiedString(pvType);

--- a/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/scheduler/OneWireUpdateScheduler.java
+++ b/bundles/binding/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/scheduler/OneWireUpdateScheduler.java
@@ -280,7 +280,7 @@ public class OneWireUpdateScheduler {
         for (int lvNumber : cvScheduleMap.keySet()) {
             List<String> lvItemListe = cvScheduleMap.get(lvNumber);
             if (lvItemListe.contains(pvItemName)) {
-                logger.debug("remove item=" + pvItemName + " from scheduler!");
+                logger.debug("remove item={} from scheduler!", pvItemName);
                 lvItemListe.remove(pvItemName);
             }
         }
@@ -311,7 +311,7 @@ public class OneWireUpdateScheduler {
                 } else {
                     logger.debug("Autorefresh: Adding {} item(s) with refresh time {} to reader queue.",
                             lvItemNameList.size(), ivAutoRefreshTimeInSecs);
-                    logger.debug("Update Task isAlive: " + ivOneWireUpdateTask.isAlive());
+                    logger.debug("Update Task isAlive: {}", ivOneWireUpdateTask.isAlive());
                     if (!ivOneWireUpdateTask.isAlive()) {
                         logger.debug("create and start a new Update Task again...");
 
@@ -326,10 +326,10 @@ public class OneWireUpdateScheduler {
                         // only add items to queue which aren't already in queue
                         for (String lvItemName : lvItemNameList) {
                             if (!ivUpdateQueue.contains(lvItemName)) {
-                                logger.debug("add item " + lvItemName + " to updateQueue");
+                                logger.debug("add item {} to updateQueue", lvItemName);
                                 ivUpdateQueue.add(lvItemName);
                             } else {
-                                logger.debug("didn't add item " + lvItemName + " to updateQueue, it is alread there");
+                                logger.debug("didn't add item {} to updateQueue; it is already there", lvItemName);
                             }
                         }
                     }

--- a/distribution/openhabhome/configurations/openhab_default.cfg
+++ b/distribution/openhabhome/configurations/openhab_default.cfg
@@ -742,6 +742,15 @@ logging:pattern=%date{ISO8601} - %-25logger: %msg%n
 # upon read (optional, defaults to 3)
 #onewire:retry=
 
+# the number of attempts that will be made to connect to the owserver
+# after a failed connection attempt. (Optional; defaults to 3)
+# If set to 0, no retries will be attempted.
+#onewire:server_retries=3
+
+# the time between connection attempts (optional, defaults to 60 seconds).
+# may not be less than 5 seconds
+#onewire:server_retryInterval=60
+
 # defines which temperature scale owserver should return temperatures in. Valid
 # values are CELSIUS, FAHRENHEIT, KELVIN, RANKINE (optional, defaults to CELSIUS).
 #onewire:tempscale=

